File: 0128-longest-consecutive-sequence/128.py
class Solution:
    def longestConsecutive(self, nums: List[int]) -> int:
        """
        Approach:
        1. Hash Set -- T.C: O(N), S.C: O(N)
             - Use set, start only from sequence starts.
        2. Sorting -- T.C: O(N log N), S.C: O(1)
             - Sort and count consecutive elements.
        """

        # Approach 1 -- Hash Set
        s = set(nums)
        longest = 0

        for num in s:
            if num - 1 not in s:
                curr = num
                streak = 1

                while curr + 1 in s:
                    curr += 1
                    streak += 1

                longest = max(longest, streak)

        return longest

        # Approach 2 -- Sorting
        # if not nums:
        #     return 0
        # nums.sort()
        # longest = curr = 1
        # for i in range(1, len(nums)):
        #     if nums[i] == nums[i-1]:
        #         continue
        #     elif nums[i] == nums[i-1] + 1:
        #         curr += 1
        #     else:
        #         longest = max(longest, curr)
        #         curr = 1
        # return max(longest, curr)

File: 0128-longest-consecutive-sequence/128.rs

use std::collections::HashSet;

impl Solution {
    pub fn longest_consecutive(nums: Vec<i32>) -> i32 {
        /*
         * Approach:
         * 1. Hash Set -- T.C: O(N), S.C: O(N)
         *      - Use set, start only from sequence starts.
         * 2. Sorting -- T.C: O(N log N), S.C: O(1)
         *      - Sort and count consecutive elements.
         */

        // Approach 1 -- Hash Set
        let set: HashSet<i32> = nums.iter().cloned().collect();
        let mut longest = 0;

        for &num in &set {
            if !set.contains(&(num - 1)) {
                let mut curr = num;
                let mut streak = 1;

                while set.contains(&(curr + 1)) {
                    curr += 1;
                    streak += 1;
                }

                longest = longest.max(streak);
            }
        }

        longest

        /* Approach 2 -- Sorting
        let mut nums = nums;
        if nums.is_empty() { return 0; }
        nums.sort();
        let mut longest = 1;
        let mut curr = 1;
        for i in 1..nums.len() {
            if nums[i] == nums[i-1] { continue; }
            else if nums[i] == nums[i-1] + 1 { curr += 1; }
            else {
                longest = longest.max(curr);
                curr = 1;
            }
        }
        longest.max(curr)
        */
    }
}
